pennylane/qnodes/base.py

```diff
@@ -474,8 +489,27 @@ def _construct(self, args, kwargs):
         """
         # pylint: disable=attribute-defined-outside-init, too-many-branches, too-many-statements
 
-        if self.arg_vars is None or self.kwarg_vars is None:
-            self.arg_vars, self.kwarg_vars = self._make_variables(args, kwargs)
+        def equal_dicts(a, b):
+            """Return True iff a and b are equal dicts."""
+            if b is None:
+                return False
+
+            if len(a) != len(b):
+                return False
+
+            for key, val in a.items():
+                if np.any(b[key] != val):
+                    return False
+            return True
```

Comment:
It might be worth to move this to `utils.py` as it doesn't seem to be strictly connected to the logic of `_construct`.

Reply:
It's specific to this use case, since it makes assumptions about the dicts (they are kwargs dicts with default values already filled in, so `a` and `b` always have the same keys for example).

Reply:
Aaand I just noted the comparison does not work with arbitrary nested structures. I'll fix it.